### PR TITLE
Fix SSL_set_ciphersuites to set even if no call to SSL_set_cipher_list

### PR DESCRIPTION
**Pull request by raja-ashok**
Currently only calling `SSL_set_ciphersuites` is not configuring TLS1.3 ciphersuite. Instead it requires additional call to `SSL_set_cipher_list` with null String or some valid TLS1.2 ciphersuite.

But this behaviour is not with the Context level API `SSL_CTX_set_ciphersuites`. 

https://www.mail-archive.com/openssl-users@openssl.org/msg86247.html

##### Checklist

- [x] tests are added or updated

### CODE REVIEW

File: test/sslapitest.c [hunk 1]
```diff
@@ -3288,6 +3288,111 @@ static int test_ciphersuite_change(void)
...
+                   SSL_CIPHER_get_name(SSL_get_current_cipher(clientssl))) != 0)
+            goto end;
+        if (strncmp(t13_cipher, SSL_CIPHER_get_name(SSL_get_current_cipher(serverssl)),
+                strlen(SSL_CIPHER_get_name(SSL_get_current_cipher(serverssl)))) != 0)
+            goto end;
+        if (strncmp(t13_cipher, SSL_CIPHER_get_name(SSL_get_current_cipher(clientssl)),
+                strlen(SSL_CIPHER_get_name(SSL_get_current_cipher(clientssl)))) != 0)
+            goto end;
+
+        SSL_free(serverssl);
```
Comment (paulidale) [inline test/sslapitest.c]: This isn't the usual error handling setup.

```
    ret = 1;
end:
    SSL_free(serverssl);
    SSL_free(clientssl);
    SSL_CTX_free(sctx);
    SSL_CTX_free(cctx);
    return ret;
```

With the addition of `int ret = 0;` at the start of the function.

File: test/sslapitest.c [hunk 2]
```diff
@@ -3288,6 +3288,111 @@ static int test_ciphersuite_change(void)
...
+                goto end;
+        }
+
+        if (!TEST_true(create_ssl_connection(serverssl, clientssl, SSL_ERROR_NONE)))
+            goto end;
+
+        if (strcmp(SSL_CIPHER_get_name(SSL_get_current_cipher(serverssl)),
+                   SSL_CIPHER_get_name(SSL_get_current_cipher(clientssl))) != 0)
+            goto end;
+        if (strncmp(t13_cipher, SSL_CIPHER_get_name(SSL_get_current_cipher(serverssl)),
```
Comment (paulidale) [inline test/sslapitest.c]: Wouldn't a basic `strcmp` be better here?

File: test/sslapitest.c [hunk 3]
```diff
@@ -3288,6 +3288,111 @@ static int test_ciphersuite_change(void)
...
+                strlen(SSL_CIPHER_get_name(SSL_get_current_cipher(serverssl)))) != 0)
+            goto end;
+        if (strncmp(t13_cipher, SSL_CIPHER_get_name(SSL_get_current_cipher(clientssl)),
```
Comment (paulidale) [inline test/sslapitest.c]: And here?

File: test/sslapitest.c [hunk 4]
```diff
@@ -3288,6 +3288,111 @@ static int test_ciphersuite_change(void)
...
+ * Test 0 = Set TLS1.3 cipher on context
+ * Test 1 = Set TLS1.3 cipher on SSL
+ * Test 2 = Set TLS1.3 and TLS1.2 cipher on context
+ * Test 3 = Set TLS1.3 and TLS1.2 cipher on SSL
+ */
+static int test_tls13_ciphersuite(int idx)
+{
+    SSL_CTX *sctx = NULL, *cctx = NULL;
+    SSL *serverssl = NULL, *clientssl = NULL;
+    const char *t13_ciphers[] = {
```
Comment (paulidale) [inline test/sslapitest.c]: `static` ?

File: test/sslapitest.c [hunk 5]
```diff
@@ -3288,6 +3288,111 @@ static int test_ciphersuite_change(void)
...
+        TLS1_3_RFC_AES_128_GCM_SHA256,
+        TLS1_3_RFC_CHACHA20_POLY1305_SHA256,
+        TLS1_3_RFC_AES_128_CCM_SHA256,
+        TLS1_3_RFC_AES_256_GCM_SHA384":"TLS1_3_RFC_CHACHA20_POLY1305_SHA256,
+        TLS1_3_RFC_AES_128_CCM_8_SHA256":"TLS1_3_RFC_AES_128_CCM_SHA256};
```
Comment (paulidale) [inline test/sslapitest.c]: Line break before the `}`.

File: test/sslapitest.c [hunk 6]
```diff
@@ -3288,6 +3288,111 @@ static int test_ciphersuite_change(void)
...
+
+    for (i = 0; i < (int)(sizeof(t13_ciphers)/sizeof(t13_ciphers[0])); i++) {
+        t13_cipher = t13_ciphers[i];
+        if (!TEST_true(create_ssl_ctx_pair(TLS_server_method(), TLS_client_method(),
+                                           TLS1_VERSION, TLS_MAX_VERSION,
+                                           &sctx, &cctx, cert, privkey)))
+            goto end;
+
+        if (set_at_ctx) {
+            if (t13_cipher && (!TEST_true(SSL_CTX_set_ciphersuites(sctx, t13_cipher))))
```
Comment (paulidale) [inline test/sslapitest.c]: Are the parenthesis around the `!TEST_true(...)` really necessary?

File: test/sslapitest.c [hunk 7]
```diff
@@ -3288,6 +3288,111 @@ static int test_ciphersuite_change(void)
...
+                goto end;
+            if (t13_cipher && (!TEST_true(SSL_CTX_set_ciphersuites(cctx, t13_cipher))))
```
Comment (paulidale) [inline test/sslapitest.c]: Maybe combine the checks?

```
if (t13_cipher) {
    if (!TEST_true(SSL_CTX_set_ciphersuites(sctx, ...))
        || !TEST_true(SSL_CTX_set_ciphersuites(cctx, ...)))
        goto end;
}
```

File: test/sslapitest.c [hunk 8]
```diff
@@ -3288,6 +3288,111 @@ static int test_ciphersuite_change(void)
...
+                goto end;
+            if (t12_cipher && (!TEST_true(SSL_CTX_set_cipher_list(sctx, t12_cipher))))
```
Comment (paulidale) [inline test/sslapitest.c]: Ditto for the parenthesis and combination.

File: test/sslapitest.c [hunk 9]
```diff
@@ -3288,6 +3288,111 @@ static int test_ciphersuite_change(void)
...
+            if (t12_cipher && (!TEST_true(SSL_CTX_set_cipher_list(cctx, t12_cipher))))
+                goto end;
+        }
+
+        if (!TEST_true(create_ssl_objects(sctx, cctx, &serverssl, &clientssl,
+                                                 NULL, NULL)))
+            goto end;
+
+        if (set_at_ssl) {
+            if (t13_cipher && (!TEST_true(SSL_set_ciphersuites(serverssl, t13_cipher))))
```
Comment (paulidale) [inline test/sslapitest.c]: Again...

File: test/sslapitest.c [hunk 10]
```diff
@@ -3288,6 +3288,111 @@ static int test_ciphersuite_change(void)
...
+                goto end;
+            if (t13_cipher && (!TEST_true(SSL_set_ciphersuites(clientssl, t13_cipher))))
+                goto end;
+            if (t12_cipher && (!TEST_true(SSL_set_cipher_list(serverssl, t12_cipher))))
```
Comment (paulidale) [inline test/sslapitest.c]: Again...

**Comment by raja-ashok** [comment]
> Why is the on 1.1.1 and not master?
> Back merging to 1.1.1 is generally preferable to forward merging to master (although it isn't necessary).

Sure. I will raise fix for master branch also.

File: test/sslapitest.c (hunk 2 above)
Comment (raja-ashok) [reply test/sslapitest.c]: For somecases, `t13_cipher` holds more than one ciphersuites. That is the reason `strncmp` is used here.

File: test/sslapitest.c (hunk 3 above)
Comment (raja-ashok) [reply test/sslapitest.c]: Here also `t13_cipher` may hold more than 1 cipher.

File: test/sslapitest.c (hunk 4 above)
Comment (raja-ashok) [reply test/sslapitest.c]: Ya, it can be changed to static also.

File: ssl/ssl_ciph.c [hunk 11]
```diff
@@ -1390,8 +1391,7 @@ int SSL_set_ciphersuites(SSL *s, const char *str)
 {
     int ret = set_ciphersuites(&(s->tls13_ciphersuites), str);
 
-    if (ret && s->cipher_list != NULL) {
-        /* We already have a cipher_list, so we need to update it */
+    if (ret) {
         return update_cipher_list(&s->cipher_list, &s->cipher_list_by_id,
                                   s->tls13_ciphersuites);
     }
```
Comment (mattcaswell) [inline ssl/ssl_ciph.c]: No {} required on a single line if

File: test/sslapitest.c [hunk 12]
```diff
@@ -3288,6 +3288,114 @@ static int test_ciphersuite_change(void)
...
+            }
+        }
+
+        if (!TEST_true(create_ssl_connection(serverssl, clientssl, SSL_ERROR_NONE)))
+            goto end;
+
+        if (strcmp(SSL_CIPHER_get_name(SSL_get_current_cipher(serverssl)),
+                SSL_CIPHER_get_name(SSL_get_current_cipher(clientssl))) != 0)
+            goto end;
+        if (strncmp(t13_cipher, SSL_CIPHER_get_name(SSL_get_current_cipher(serverssl)),
```
Comment (mattcaswell) [inline test/sslapitest.c]: `TEST_strn_eq` should be used here instead

File: test/sslapitest.c [hunk 13]
```diff
@@ -3288,6 +3288,114 @@ static int test_ciphersuite_change(void)
...
+                if (!TEST_true(SSL_set_cipher_list(serverssl, t12_cipher))
+                    || !TEST_true(SSL_set_cipher_list(clientssl, t12_cipher)))
+                    goto end;
+            }
+        }
+
+        if (!TEST_true(create_ssl_connection(serverssl, clientssl, SSL_ERROR_NONE)))
+            goto end;
+
+        if (strcmp(SSL_CIPHER_get_name(SSL_get_current_cipher(serverssl)),
```
Comment (mattcaswell) [inline test/sslapitest.c]: `TEST_str_eq` should be used here instead.

File: ssl/ssl_ciph.c (hunk 11 above)
Comment (raja-ashok) [reply ssl/ssl_ciph.c]: Done

File: test/sslapitest.c (hunk 12 above)
Comment (raja-ashok) [reply test/sslapitest.c]: Done

File: test/sslapitest.c (hunk 13 above)
Comment (raja-ashok) [reply test/sslapitest.c]: Done

**Comment by mattcaswell** [comment]
Replaced by #9079.

**Pull request closed without merging**